### PR TITLE
Fixes #27393 - Update redux for redux-hooks

### PR DESCRIPTION
**Pull request by amirfefer**

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js [hunk 1]
```diff
@@ -38,11 +37,6 @@ describe('bookmark form', () => {
       query: '',
     });
   });
-  it('should render the entire form', () => {
-    const { wrapper } = setup();
-
-    expect(toJson(wrapper)).toMatchSnapshot();
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/form/form.test.js:44]: this creates a very big snap because it adds redux related props

File: webpack/assets/javascripts/react_app/components/FactCharts/__test__/FactChart.test.js [hunk 2]
```diff
@@ -22,12 +22,4 @@ describe('factCharts', () => {
     );
     expect(toJson(wrapper)).toMatchSnapshot();
   });
-
-  it('should open modal after click', () => {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/FactCharts/__test__/FactChart.test.js:26]: This test doesn't test anything, because `FactChart` component doesn't receive `onClick` at all.
This test is broken after upgrading redux.

**Comment by tbrisker** [comment]
Does this require a bump to foreman-vendor as well to get a newer redux?

**Comment by amirfefer** [comment]
@tbrisker, yes - already exist :) 
https://github.com/theforeman/foreman-js/pull/27

**Comment by tbrisker** [comment]
so we need to update the version in package.json to match the version of vendor that contains this change, and bump the version in packaging as well

File: package.json [hunk 3]
```diff
@@ -20,7 +20,7 @@
     "create-react-component": "yo react-domain"
   },
   "dependencies": {
-    "@theforeman/vendor": "^0.1.1",
+    "@theforeman/vendor": "1.4.1-update-redux.0",
```
Comment (amirfefer) [inline package.json]: This is a temporary version for testing/reviewing - please do not merge with this version

**Comment by amirfefer** [comment]
@sharvit - travis is green :) other failures seem unrelated...
can you please review ?

**Comment by sharvit** [comment]
> @sharvit - travis is green :) other failures seem unrelated...
> can you please review ?

Thanks @amirfefer, code-wise LGTM :+1: 
Are you sure the Jenkins tests are unrelated? They look suspicious...

**Comment by amirfefer** [comment]
[test foreman]

**Comment by sharvit** [comment]
Are the test related? It feels like we don't have a clear answer

**Comment by amirfefer** [comment]
It seems not related to me (after I ran it again we got different failures).

**Comment by tbrisker** [comment]
test failures are not related. 
Would merging this cause tests to fail for plugins using redux due to mismatching snapshots?
BTW, this PR shows a big downside of relying mostly on snapshots for testing - every little change in a dependency requires significant changes to the snapshots, which compounds even more when plugins use snapshot tests. i.e., even though there was no breaking change in functionality, this could cause test failures because of different html output rendered.

**Comment by sharvit** [comment]
> Would merging this cause tests to fail for plugins using redux due to mismatching snapshots?

Yes, the new `redux` version introduced here contains breaking changes.


> BTW, this PR shows a big downside of relying mostly on snapshots for testing - every little change in a dependency requires significant changes to the snapshots, which compounds even more when plugins use snapshot tests. i.e., even though there was no breaking change in functionality, this could cause test failures because of different html output rendered.

It's because we still have technical debts, notice all the failures are only when we use to "old" folder structure and not using the `react-redux-test-utils`.
Unreliable tests were one of the reasons we decided to migrate.

**Comment by amirfefer** [comment]
> this PR shows a big downside of relying mostly on snapshots for testing - every little change in a dependency requires significant changes to the snapshots...

This PR shows us that we used snapshots incorrectly with old connected components

**Comment by amirfefer** [comment]
@sharvit - can you please release a new version of `foreman-vendor` so we can merge this?

**Comment by Ron-Lavi** [comment]
Thanks, @amirfefer,
While trying to test it on `centos-luna-devel` machine, 
the application failed on start:
```
Uncaught Error: It looks like you are passing several store enhancers to createStore(). This is not supported. Instead, compose them together to a single function.
    at createStore (redux.js:66)
    at generateStore (index.js:22)
   ...
```
this code in `.../redux/index.js` need to be changed into:
```
const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const generateStore = () =>
  createStore(
    reducers,
    composeEnhancer(applyMiddleware(thunk)),
  );
```

After fixing that, everything worked well. Found no regression

**Comment by jeremylenz** [comment]
Would this be a good place to add [eslint-plugin-react-hooks](https://reactjs.org/docs/hooks-rules.html#eslint-plugin)?

(Not that I'm trying to delay merging :smile:)

**Comment by sharvit** [comment]
> Would this be a good place to add [eslint-plugin-react-hooks](https://reactjs.org/docs/hooks-rules.html#eslint-plugin)?
> 
> (Not that I'm trying to delay merging )

Can we donate it to the patternfly-react linter?
https://github.com/patternfly/patternfly-react/tree/master/packages/eslint-plugin-patternfly-react

**Comment by jeremylenz** [comment]
@sharvit Since we use patternfly-react's eslint plugin anyway, it would have the same effect for Foreman as if we just added it to our own `.eslintrc`.  But we would also be forcing the rule on anyone else who uses patternfly-react.  Personally I would be fine with that.. thoughts?

**Comment by sharvit** [comment]
> @sharvit Since we use patternfly-react's eslint plugin anyway, it would have the same effect for Foreman as if we just added it to our own `.eslintrc`. But we would also be forcing the rule on anyone else who uses patternfly-react. Personally I would be fine with that.. thoughts?

Yes, it would have the same effect.
I don't see it as "forcing the rule on anyone else" but as "collaborate with other projects so we can decide and maintain a set of tools together". Therefore I would prefer donating to pf in this case.

**Comment by amirfefer** [comment]
Updated vendor to version 2.0.0 
@sharvit about the eslint plugin - I opened a PR for it in foreman, and i'll open it in patternfly too, but until it will be added to patternfly let's add it to foreman

**Comment by sharvit** [comment]
> Updated vendor to version 2.0.0
> @sharvit about the eslint plugin - I opened a PR for it in foreman, and i'll open it in patternfly too, but until it will be added to patternfly let's add it to foreman

Thanks, make sense.

Notice the error in travis

File: webpack/assets/javascripts/react_app/components/bookmarks/bookmarks.test.js [hunk 4]
```diff
@@ -91,7 +91,7 @@ describe('bookmarks loading', () => {
 
 describe('bookmarks', () => {
   it('empty state', () => {
-    expect(toJson(render(setup().component))).toMatchSnapshot();
+    expect(toJson(mount(setup().component).find('Dropdown'))).toMatchSnapshot();
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/bookmarks/bookmarks.test.js:94]: using `render` triggers a warning: 
```
react-dom-server.browser.development.js:104 Warning: useLayoutEffect does nothing on the server,
because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch 
between the initial, non-hydrated UI and the intended UI.
```
which cause by new implementation of `connect`:
https://github.com/reduxjs/react-redux/issues/1373

File: webpack/assets/javascripts/react_app/components/bookmarks/index.js [hunk 5]
```diff
@@ -90,7 +90,7 @@ class BookmarkContainer extends React.Component {
                 />
               ))) || <MenuItem disabled> {__('None found')}</MenuItem>)}
           {status === STATUS.ERROR && (
-            <MenuItem key="bookmarks-errors">
+            <MenuItem bsClass="bookmarks-errors" key="bookmarks-errors">
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/bookmarks/index.js]: How is this related?

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/bookmarks/index.js]: it was, for reducing snapshots and getting only the related part for it, but I have found another way. 
fixed ! :)

**Review by sharvit — APPROVED** [review]
Thanks @amirfefer, looks good and works as expected.

File: webpack/assets/javascripts/react_app/components/bookmarks/bookmarks.test.js (hunk 4 above)
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/bookmarks/bookmarks.test.js:94]: We're not using server rendering, are we?

**Review by tbrisker — COMMENTED** [review]
This needs updates to packaging for the new vendor version. Do you think it would be done in time for 1.24? we are just two weeks away from stabilization and updating vendor version now means we would need to update all plugins to it and make sure they are still working properly with the new version.

**Comment by sharvit** [comment]
> This needs updates to packaging for the new vendor version. Do you think it would be done in time for 1.24? we are just two weeks away from stabilization and updating vendor version now means we would need to update all plugins to it and make sure they are still working properly with the new version.

IMO we should not merge it and wait for `1.25`

**Comment by jeremylenz** [comment]
@sharvit @amirfefer Opened PR for patternfly https://github.com/patternfly/patternfly-react/pull/3083

**Comment by jeremylenz** [comment]
> Updated vendor to version 2.0.0
> @sharvit about the eslint plugin - I opened a PR for it in foreman, and i'll open it in patternfly too, but until it will be added to patternfly let's add it to foreman

@amirfefer sorry I didn't see that you were also going to open in patternfly, hope I didn't duplicate anything!

**Comment by amirfefer** [comment]
looks like this eslint plugin won't be merged to patternfly :(

**Comment by amirfefer** [comment]
tests failures look unrelated

**Review by xprazak2 — APPROVED** [review]
Let's get this in. Thanks everyone!

**Comment by xprazak2** [comment]
Needs packaging PR, though.

**Comment by xprazak2** [comment]
Packaging: https://github.com/theforeman/foreman-packaging/pull/4263

**Review by sharvit — APPROVED** [review]
IMO it is ready to be merged, waiting for the packaging PR to get merged first: https://github.com/theforeman/foreman-packaging/pull/4302

**Comment by sharvit** [comment]
Another packaging PR with vendor v3 https://github.com/theforeman/foreman-packaging/pull/4330

**Comment by sharvit** [comment]
> Another packaging PR with vendor v3 [theforeman/foreman-packaging#4330](https://github.com/theforeman/foreman-packaging/pull/4330)

v3.1 - https://github.com/theforeman/foreman-packaging/pull/4348

**Review by evgeni — APPROVED** [review]
packaging-wise: ack (but should be merged togther with #7164 and #7178 as we don't have vendor 2.0 packaged/PR), I have no idea about JS tho ;)

**Comment by sharvit** [comment]
Closing, merged here: #7178
